### PR TITLE
Refactor TopInstructions

### PR DESCRIPTION
**Pull request by maureensturgeon**
This may be easier to review if you look at the first two commits independently.

**First commit:** pulled out data requests into `topInstructionsHelpers`, added firehose helper function, refactored confusing logic, in render deconstructed `this.props` and `this.state` to make render more readable 
**Second commit:** Pulled out instructions header into a new component called `TopInstructionsHeader`, refactored unit tests

There are no functional changes included in this PR, purely refactor.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by maureensturgeon — COMMENTED** [review]
Actually, hold on this review, I've decided to do further, more drastic refactor.

**Review by maureensturgeon — COMMENTED** [review]
Ok - now it's ready for review!

File: apps/src/templates/instructions/TopInstructionsHeader.jsx [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+import {TabType} from './TopInstructions';
+import i18n from '@cdo/locale';
+import color from '../../util/color';
+import styleConstants from '../../styleConstants';
+
+const styles = {
+  paneHeaderOverride: {
+    color: color.default_text
+  },
+  audioStyleRTL: {
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructionsHeader.jsx]: NIT: including "style" in this name is redundant and doesn't match the naming of the rest of the styles

File: apps/src/templates/instructions/topInstructionsHelpers.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import $ from 'jquery';
```
Comment (jamescodeorg) [inline apps/src/templates/instructions/topInstructionsHelpers.js]: A few thoughts on this:
1) It would be nice if these methods could be easily recognizable as server calls when they're used.  One possibility night be to always reference them as e.g. `serverApi.getRubric()`.
2) Along the same lines, perhaps we can name this file something different?  ("Helpers" seems a bit vague and I'm worried we'll start putting other stuff in here that reduces clarity.)
3) I'm wondering if we want to abstract a bit more here and have these methods return a promise that returns something a bit more typed (e.g. returning an array of feedback vs. just the raw request).  I think this is a more natural abstraction and also has the benefit of improving unit testing for consuming code (because you can mock the server call more completely).  This probably needs some more thinking and can be a separate improvement.

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 3]
```diff
@@ -619,222 +559,115 @@ class TopInstructions extends Component {
...
           >
-            <div
-              ref={ref => {
-                if (ref) {
-                  this.instructions = ref;
-                }
-              }}
-            >
-              {this.props.hasContainedLevels && (
+            <div ref={ref => this.setInstructionsRef(ref)}>
```
Comment (jamescodeorg) [inline apps/src/templates/instructions/TopInstructions.jsx:663]: Maybe create a renderInstructions() function that returns one of the three possible cases of instructions?

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 4]
```diff
@@ -619,222 +559,115 @@ class TopInstructions extends Component {
...
+          handleTeacherOnlyTabClick={this.handleTeacherOnlyTabClick}
+          handleClickCollapser={this.handleClickCollapser}
+        />
+        <div style={[isCollapsed && isCSDorCSP && commonStyles.hidden]}>
           <div
             style={[
               isCSF &&
-              !this.props.hasContainedLevels &&
-              this.state.tabSelected === TabType.INSTRUCTIONS
+              !hasContainedLevels &&
```
Comment (jamescodeorg) [inline apps/src/templates/instructions/TopInstructions.jsx]: I'm finding it hard to parse this logic that selects the style inline.  Maybe we can calculate this on a separate line?

**Review by jamescodeorg — APPROVED** [review]
Phew, what a beast.

As a future improvement, it'd be nice if we could extract (or find) a tab-navigation class so we can remove all of the navigation logic.  Maybe something that the component library team can help with?

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 5]
```diff
@@ -216,7 +146,7 @@ class TopInstructions extends Component {
     const studentId = queryString.parse(window.location.search).user_id;
 
     const teacherViewingStudentWork =
-      this.props.viewAs === ViewType.Teacher &&
+      this.isViewingAsTeacher() &&
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.jsx]: love this. i'm a big fan of wrapping a line or two of code in a helper function to improve readability!

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 6]
```diff
@@ -466,16 +387,22 @@ class TopInstructions extends Component {
     win.focus();
   };
 
+  recordTopInstructionsEvent(eventName, additionalData = {}) {
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.jsx]: NIT: including "TopInstructions" in this name seems redundant in this context.

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 7]
```diff
@@ -531,59 +451,86 @@ class TopInstructions extends Component {
...
-          headers: {'X-CSRF-Token': this.state.token}
-        });
+        incrementVisitCount(latestFeedback.id, this.state.token);
       }
     },
     5000,
     {leading: true}
   );
 
+  isViewingAsStudent = () => {
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.jsx]: these one-liners as arrow functions isn't a big concern from a performance standpoint, but i still think we should be consistent about never using arrow functions as class properties.

i realize it would feel like overkill to bind them, so why don't you just set them as values in the constructor? `this.isViewingAsTeacher = props.viewAs === ViewType.Student`, then you can just check the value instead of calling a function.

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 8]
```diff
@@ -531,59 +451,86 @@ class TopInstructions extends Component {
...
+    return this.props.viewAs === ViewType.Student;
+  };
+
+  isViewingAsTeacher = () => {
+    return this.props.viewAs === ViewType.Teacher;
+  };
+
+  setInstructionsRef(ref) {
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.jsx:476]: NIT: personally i prefer doing this inline. imo it's clearer and more concise.

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 9]
```diff
@@ -531,59 +451,86 @@ class TopInstructions extends Component {
...
+    if (ref) {
+      this.instructions = ref;
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.jsx:478]: can you declare this by setting it to null right under the constructor? i'm pretty opposed to setting properties that haven't been declared anywhere.

File: apps/src/templates/instructions/TopInstructions.jsx (hunk 7 above)
Comment (cforkish) [reply apps/src/templates/instructions/TopInstructions.jsx]: actually it looks like they don't need t be bound anyway so you could just convert them to regular functions, but imo setting them as props in the constructor makes more sense. though if you expect the `viewAs` prop to change, leaving them as functions is simpler than updating their values in `componentDidUpdate`

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 10]
```diff
@@ -619,222 +559,115 @@ class TopInstructions extends Component {
...
+                user={user}
+                visible={tabSelected === TabType.COMMENTS}
                 displayKeyConcept={displayKeyConcept}
                 disabledMode={
-                  this.props.viewAs === ViewType.Student ||
-                  !this.state.teacherViewingStudentWork
+                  this.isViewingAsStudent() || !teacherViewingStudentWork
                 }
-                rubric={this.state.rubric}
+                rubric={rubric}
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.jsx:682]: there are a few more refs to declare in this chunk of code.

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 11]
```diff
@@ -619,222 +559,115 @@ class TopInstructions extends Component {
...
+                  <Instructions
+                    ref={ref => this.setInstructionsRef(ref)}
+                    longInstructions={longInstructions}
+                    onResize={this.adjustMaxNeededHeight}
+                    inTopPane
+                  />
                 )}
             </div>
-            {this.state.tabSelected === TabType.RESOURCES && (
+            {tabSelected === TabType.RESOURCES && (
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.jsx:666]: another ref to declare

File: apps/src/templates/instructions/topInstructionsHelpers.js (hunk 2 above)
Comment (maureensturgeon) [reply apps/src/templates/instructions/topInstructionsHelpers.js]: I've renamed `topInstructionsHelpers` to `topInstructionsDataApi` and imported the functions under that name so they're now referenced like `topInstructionsDataApi.getRubric()`. Let me know what you think, happy to rename again

File: apps/src/templates/instructions/TopInstructionsHeader.jsx [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+    wrapper: {
+      float: 'left'
+    },
+    button: {
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructionsHeader.jsx:23]: there are a lot of duplicated styles in here. if you add `audioStyle.RTLWrapper` you can remove `audioStyleRTL` entirely. i see a lot more duplication that can be removed but you get the idea.

File: apps/src/templates/instructions/TopInstructionsHeader.jsx [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+            iconClass="fa fa-book"
+            label={i18n.documentation()}
+            isRtl={isRtl}
+            headerHasFocus={false}
+            onClick={handleDocumentationClick}
+            isMinecraft={isMinecraft}
+          />
+        )}
+        <div style={isRtl ? styles.helpTabsRtl : styles.helpTabs}>
+          <InstructionsTab
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructionsHeader.jsx:139]: if you create a `renderInstructionsTab` helper i think you could seriously cut down on the number of lines in this component.

**Review by cforkish — CHANGES_REQUESTED** [review]
i really appreciate you taking the initiative to clean this up. overall looks like a great improvement. a few minor changes requested.

File: apps/src/templates/instructions/TopInstructions.jsx (hunk 10 above)
Comment (maureensturgeon) [reply apps/src/templates/instructions/TopInstructions.jsx:682]: I don't understand what you mean

Comment (cforkish) [reply apps/src/templates/instructions/TopInstructions.jsx:682]: can you declare `helpTab`, `commentTab`,  and `teacherOnlyTab` like you did with `instructions`?

Comment (maureensturgeon) [reply apps/src/templates/instructions/TopInstructions.jsx:682]: Oh - got it, you mean initialize them as null.

**Review by cforkish — APPROVED** [review]
LGTM!
